pkg/azdiskschedulerextender/azdiskschedulerextender.go

```diff
@@ -183,38 +186,17 @@ func filter(context context.Context, schedulerExtenderArgs schedulerapi.Extender
 	// filter the nodes based on AzDiverNode status and number of disk attachments
 	failedNodes = make(map[string]string)
 	for _, node := range allNodes {
-		ready, found := nodeNameToAzDriverNodeStatusMap[node.Name]
-		excludeNode := false
-		if found && ready {
-			// TODO: remove maxDiskCount calculations when we fix #1097 in the controller
-			// get AzVolumeAttachments for the node
-			azVolumeAttachments, err := getAdditionalAzVolumeAttachmentsOnNode(node.Name, diskNamesForRequestedVolumes)
-			klog.V(2).Infof(
-				"number of additional volume attachments for node %s and pod %s is %d. Number of requested volumes is %d. Err: %v",
-				node.Name, schedulerExtenderArgs.Pod.Name, len(azVolumeAttachments), len(diskRequestedByPod), err)
-			if err != nil {
-				klog.Warningf("failed to get azVolumeAttachments attached to node (%s): %v", node.Name, err)
-				continue
-			}
-
-			// get the max disk count for the node type
-			maxDiskCount, err := azureutils.GetNodeMaxDiskCountWithLabels(node.Labels)
-			if err != nil {
-				klog.Warningf("failed to get node (%s)'s max disk count: %v", node.Name, err)
-				continue
-			}
-
-			if len(azVolumeAttachments)+totalRequestedDiskCount > maxDiskCount {
-				klog.Warningf("max number of %v AzVolumeAttachments for node %s has been reached. The node will be skipped", maxDiskCount, node.Name, azDriverNodesMeta.err)
-				excludeNode = true
-				failedNodes[node.Name] = "node(s) reached the maximum number of disk attachments"
-			}
-			if !excludeNode {
+		if ready, found := nodeNameToAzDriverNodeStatusMap[node.Name]; found {
+			if ready {
 				filteredNodes = append(filteredNodes, node)
 				storeFilteredNodeNames = append(storeFilteredNodeNames, node.Name)
+			} else {
+				failedNodes[node.Name] = fmt.Sprintf("AzDriverNode for %s is not ready.", node.Name)
 			}
 		} else {
-			failedNodes[node.Name] = fmt.Sprintf("AzDriverNode for %s is not ready.", node.Name)
+			// if AzDriverNode cannot be found for the node, either 1) AzDriverNode cache is stale or 2) the node plugin has not started and needs to be scheduled.
+			// so add the node to filtered node list.
+			storeFilteredNodeNames = append(storeFilteredNodeNames, node.Name)
```

Comment:
This only works as long as the assumption that all nodes in the cluster will have AzDriverNode created for them at one point. Because otherwise we could schedule a pod on a node without the driver and it will fail provisioning disks. This is ok for now because the strategy of the extender has really shifted to only work for failover, but this is something we should watch out for in the future in case our approach changes or there are some node affinity rules that come into play.

Reply:
It is required for our driver to create `AzDriverNode` object for all nodes in cluster in order for our controller to function and a lot of logic in the controller, including `node controller` and `node availability controller` rely on this premise. We deploy our node controller as a Daemonset; this will ensure that all nodes in cluster have its own node controller running, which will also create an AzDriverNode object for its host node.